exist-core/src/main/java/org/exist/xquery/functions/map/MapType.java

```diff
@@ -1,169 +1,252 @@
         }
         return seq;
     }
 
     public AbstractMapType remove(final AtomicValue[] keysAtomicValues) {
-        IPersistentMap<AtomicValue, Sequence> tempmap = this.map;
+
+        // create a transient map
+        IMap<AtomicValue, Sequence> newMap = map.linear();
+
         for (final AtomicValue key: keysAtomicValues) {
-            if (!tempmap.containsKey(key)) { continue; }
-            tempmap = tempmap.without(key);
+            newMap = newMap.remove(key);
         }
-        return new MapType(this.context, tempmap, type);
+
+        // return an immutable map
+        return new MapType(context, newMap.forked(), type);
     }
 
     @Override
     public int size() {
-        return map.count();
+        return (int)map.size();
     }
 
     @Override
-    public Iterator<Map.Entry<AtomicValue, Sequence>> iterator() {
+    public Iterator<IEntry<AtomicValue, Sequence>> iterator() {
         return map.iterator();
     }
 
     @Override
-    public AtomicValue getKey() {
-        if (map.count() == 0)
-            {return null;}
-        final Iterator<Map.Entry<AtomicValue,Sequence>> iter = this.map.iterator();
-        return iter.next().getKey();
-    }
+    public AtomicValue key() {
+        if (map.size() > 0) {
+            final IEntry<AtomicValue, Sequence> entry = map.nth(0);
+            if (entry != null) {
+                return entry.key();
+            }
+        }
 
-    @Override
-    public Sequence getValue() {
-        return mapToSequence(this.map);
+        return null;
     }
 
-    /**
-     * Get a Sequence from an internal map representation
-     */
-    private Sequence mapToSequence(final IPersistentMap<AtomicValue, Sequence> map) {
-        if (map.count() == 0) {
-            return null;
+    @Override
+    public Sequence value() {
+        if (map.size() > 0) {
+            final IEntry<AtomicValue, Sequence> entry = map.nth(0);
+            if (entry != null) {
+                return entry.value();
+            }
         }
-        final Iterator<Map.Entry<AtomicValue,Sequence>> iter = map.iterator();
-        return iter.next().getValue();
+
+        return null;
     }
 
-    private void setKeyType(int newType) {
-        if (type == Type.ANY_TYPE)
-            {type = newType;}
+    private void setKeyType(final int newType) {
+        if (type == Type.ANY_TYPE) {
+            type = newType;
+        }
         else if (type != newType) {
             type = Type.ITEM;
-            if (map instanceof PersistentHashMap) {
-                try {
-                    PersistentTreeMap tmap = PersistentTreeMap.create(getComparator(null), null);
-                    for (final Map.Entry<AtomicValue, Sequence> entry : map) {
-                        tmap = tmap.assoc(entry.getKey(), entry.getValue());
-                    }
-                    map = tmap;
-                } catch (final XPathException e) {
-                }
-            }
+//            try {
```

Comment:
Is this commented out code necessary? If not, please remove it.

Reply:
I have now removed it.